File: ch02/src/problems/HornersRule.java
package problems;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class HornersRule {
	
	public static double hornersRule(double[] coeffs, double x) {
		double y = 0;
		for (int i = coeffs.length - 1; i >= 0; i--)
			y = coeffs[i] + x * y;
		return y;
	}
	
	public static double[] inputCoeffs() {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		
		double[] coeffs;
		int n = 0;
		
		System.out.println("Please input the degree of the polynomial (n): ");
		try {
			n = Integer.parseInt(br.readLine()) + 1;
		} catch(NumberFormatException | IOException nfe) {
			System.err.println("Invalid Format!");
		}
		
		coeffs = new double[n];
		
		System.out.format("Please enter the coefficients in order a0, a1, ..., a%d: ", n);
		for (int coeff = 0; coeff < n; coeff++) {
			try {
				coeffs[coeff] = Double.parseDouble(br.readLine());
			} catch(NumberFormatException | IOException nfe) {
				System.err.println("Invalid Format!");
			}
		}
		
		return coeffs;
	}
	
	public static void main(String[] args) {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		double[] coeffs;
		double x = 0.0;
		
		System.out.println("This is Horner's rule for evaluating a polynomial.");
		coeffs = inputCoeffs();
		System.out.print("Please enter a value for x: ");
		try {
			x = Double.parseDouble(br.readLine());
		} catch(NumberFormatException | IOException nfe) {
			System.err.println("Invalid Format!");
		}
		
		System.out.format("For f(x) = %f", coeffs[0]);
		for (int i = 1; i < coeffs.length; i++)
			System.out.format((coeffs[i] > 0.0 ? " + " : " ") + "%fx^%d", coeffs[i], i);
		System.out.format(",\nResult f(%f) = %f", x, hornersRule(coeffs, x));
	}
}

File: cpp/src/algorithms/polynomials/horners_rule.cpp
// Horners-Rule algorithm from p.2-3 on pg. 41 of CLRS 3e

#include <exception>
#include <iostream>
#include <string>

struct io_error : std::runtime_error {
  io_error(const std::string &message = "I/O error")
      : std::runtime_error(message) {}
};

double hornersRule(double coeffs[], int n, double x);
double *inputCoeffs(int &n);

int main(int argc, char *argv[]) {
  double x = 0.0;
  int len = 0;

  std::cout << "This is Horner's rule for evaluating a polynomial. ";

  double *coeffs = inputCoeffs(len);

  std::cout << "Please enter a value for x: ";
  try {
    std::cin >> x;
  } catch (io_error &e) {
    std::cout << "Cannot read input.";
  }

  std::cout << "For f(x) = " << coeffs[0];
  for (int i = 1; i <= len; i++)
    std::cout << (coeffs[i] > 0.0 ? " + " : " ") << "x"
              << (i > 1 ? "^" + std::to_string(i) : "");
  std::cout << ",\nResult: f(" << x << ") = " << hornersRule(coeffs, len, x)
            << '\n';

  return EXIT_SUCCESS;
}

double hornersRule(double coeffs[], int n, double x) {
  double y = 0;

  for (int i = n; i >= 0; i--)
    y = coeffs[i] + x * y;

  return y;
}

double *inputCoeffs(int &n) {
  double *coeffs;
  n = 0;

  std::cout << "Please input the degree of the polynomial (n): ";
  try {
    std::cin >> n;
  } catch (io_error) {
    std::cout << "Invalid Format!";
  }

  coeffs = new double[n];

  std::cout << "Please enter the coefficients in order a0, a1, ..., a" << n
            << ": \n";
  for (int coeff = 0; coeff <= n; coeff++) {
    try {
      std::cin >> coeffs[coeff];
    } catch (io_error) {
      std::cout << "Invalid Format!";
    }
  }

  return coeffs;
}
